mlflow/pyfunc/__init__.py

```diff
@@ -438,26 +492,45 @@ def spark_udf(spark, model_uri, result_type="double"):
 
     def predict(*args):
         model = SparkModelCache.get_or_load(archive_path)
-        schema = {str(i): arg for i, arg in enumerate(args)}
+        input_schema = model.metadata.get_input_schema()
         pdf = None
+
+        def convert_type(series: pandas.Series, type_: np.dtype):
+            if series.dtype == type_:
+                return series
+            return series.astype(type_, errors="ignore")
+
         for x in args:
             if type(x) == pandas.DataFrame:
                 if len(args) != 1:
                     raise Exception("If passing a StructType column, there should be only one "
                                     "input column, but got %d" % len(args))
                 pdf = x
         if pdf is None:
-            # Explicitly pass order of columns to avoid lexicographic ordering (i.e., 10 < 2)
-            columns = [str(i) for i, _ in enumerate(args)]
-            pdf = pandas.DataFrame(schema, columns=columns)
+            if input_schema is None:
+                data = dict([(str(i), x) for i, x in enumerate(args)])
+                # Explicitly pass order of columns to avoid lexicographic ordering (i.e., 10 < 2)
+                pdf = pandas.DataFrame(data=data, columns=[str(i) for i in range(len(args))])
+            else:
+                if len(args) != len(input_schema.columns):
```

Comment:
To confirm, are we sure we want to throw if there are extra columns in the input? e.g. should we worry about the potential use case where you're applying a model as part of a pipeline & there are some extra columns that are unused by the current model but are used further down the pipeline?

Reply:
Same goes for the serving use case, e.g. the protobuf convention is to ignore unknown fields in REST API input, so maybe we should do the same?

Reply:
Hmm I did not get the pipeline bit - these are the args passed to the model itself, the subsequent pipeline elements can still receive additional stuff?

As for whether we should follow protobuf convention - that's a good question. Maybe we should.  Also, I've realized this is a similar question to whether we should raise exception if there are incompatible types and we should be consistent in handling it.